README
SBLIM CIM Client for Java
==========================

(C) Copyright IBM Corp. 2005, 2009

 THIS FILE IS PROVIDED UNDER THE TERMS OF THE ECLIPSE PUBLIC LICENSE
 ("AGREEMENT"). ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS FILE
 CONSTITUTES RECIPIENTS ACCEPTANCE OF THE AGREEMENT.

 You can obtain a current copy of the Eclipse Public License from
 http://www.opensource.org/licenses/eclipse-1.0.php

Author:             Viktor Mihajlovski <mihajlov@de.ibm.com>
Contributors:
                    Thorsten Schaefer  <thschaef@de.ibm.com>
                    Alexander Wolf-Reber <a.wolf-reber@de.ibm.com>
                    Ramandeep S Arora <arorar@us.ibm.com>

Last Change Date: 2009-03-10


About this Package
------------------
The purpose of this package is to provide a CIM Client Class Library
for Java applications. It complies to the DMTF standard CIM Operations over
HTTP and with JCP JSR48. A RFC 2614 compliant SLP package is included for WBEM
service discovery.

To learn more about DMTF visit http://www.dmtf.org.
More info about the Java Community Process and JSR48 can be found at
http://www.jcp.org and http://www.jcp.org/en/jsr/detail?id=48.

To learn more about the SBLIM CIM Client for Java download and extract the 
documentation archive. You'll find a lot of useful information on the javadoc
overview page.

Requirements
------------
In order to build, install and operate this package you will need a 
Java SDK or Run Time Environment of version 1.5 or higher, depending on
whether you build binaries from sources or use the pre-built binaries (JAR
file).

As a requirement for building the SBLIM CIM Client, Apache Ant 
(http://ant.apache.org) needs to be setup properly. We would like to recommend
a Apache Ant version 1.6.5 or higher. (Lower versions might also work properly
here)


Building and Installing the CIM Client
--------------------------------------
The build process is based on Apache Ant, so it is very easy to build the 
SBLIM CIM Client.

Running Sun Java SDK:
If you have a xerces parser included into your systems CLASSPATH environment 
variable, no further dependencies need to be resolved.
If you don't have a xerces parser included, please create a directory named 
'lib' in the 'basedir' directory of the SBLIM CIM Client. Please copy the 
'xerces' Java Archives (JAR files) into this 'lib' directory.

An Ant build.xml file is in place which supports the following targets:
  
ant clean         - Cleans up the directory structure
ant build_core    - Builds the SBLIM CIM Client core classes
ant build_utst    - Builds the SBLIM CIM Client unit tests
ant build_smpl    - Builds the SBLIM CIM Client samples
ant package       - Builds the JAR files
ant java-doc      - Composes the JavaDoc for the SBLIM CIM Client
ant release       - Builds release TAR balls and Zip archives
ant unittest      - Executes the unit tests
ant all           - Build release archives and executes the unit tests
ant cvs-update    - Performs a 'cvs update -AdP' on the project
ant help          - Displays a help message

The execution of the build process is very easy. Once, Apache Ant is setup 
properly on your system, just execute 'ant <target>' in the 'basedir' directory
of the SBLIM CIM Client. The output will be written into a new created
directory named 'build'. That's all.


Using the SBLIM CIM Client
--------------------------
The SBLIM CIM Client is a Java implementation that can be used in all 
applications which deal with CIMOMs.
It is packaged into a Java Archive (JAR file) that can be consumed by other
Java applications as needed.

sblim-cim-client2.jar:	This jar contains the core client with the SBLIM API

Additionally, the SBLIM CIM Client is also packaged as RPM package which can be
used in Linux Distributions to be installed. The RPM is based on the 
infrastructure that is defined by the JPackage project.


NEWS
Version 2.2.5
================
   2717 Update detailed release history HTML for 2.2.5
   2716 Sync up javax.* javadoc with JSR48 1.0.0 Final V
   2719 TCK: CIM APIs should not generate NullPointerException
   2594 CR28: Support CIMErrorDescription HTTP field
   2718 Bad CIMStatusCode generates NumberFormatException
   2538 CR14: Support new CORRELATOR element
   2715 Add VALUE.NULL support
   2714 Add detailed CIM-XML parser test based on DSP0201 2.4
   2713 Enforce loose validation of CIM-XML documents
   2712 SimpleReqNode allows any CIM element as child
   2709 Lower the level of the EOF message to FINE
   2711 LOCALNAMESPACEPATH allows 0 NAMESPACE children
   2710 parseVALUEOBJECTWITH(LOCAL)PATH ignores (LOCAL)CLASSPATH child
   2708 CIMNode quietly ignores DECLARATION child
   2707 INSTANCENAME ignores KEYVALUE and VALUE.REFERENCE children
   2706 Bad PARAMETER.REFARRAY ARRAYSIZE generates NumberFormatException
   2705 PARAMETER.ARRAY does not require TYPE attribute
   2704 PARAMETER does not require TYPE attribute
   2703 MethodNode should not require TYPE attribute
   2702 Bad PROPERTY.ARRAY ARRAYSIZE generates NumberFormatException
   2701 PROPERTY.ARRAY does not require TYPE attribute
   2700 PROPERTY does not require TYPE attribute
   2699 parseQUALIFIER does not require TYPE attribute
   2697 (I)MethodResponseNode allows ERROR with PARAMVALUE
   2696 parseIRETURNVALUE ignores VALUE and VALUE.ARRAY
   2695 parseMETHODCALL allows LOCALCLASSPATH and LOCALINSTANCEPATH
   2694 NAME attribute not required by DOM parser (part 2)
   2693 ReturnValueNode allows invalid PARAMTYPE attribute
   2691 RETURNVALUE should not require PARAMTYPE attribute
   2537 Add new data types for PARAMVALUE
   2690 Remove RESPONSEDESTINATION support
   2689 createMETHODCALL should not add PARAMTYPE attribute
   2688 parseMETHODCALL looks for CIMName attribute instead of NAME
   2687 ExpParamValueNode allows VALUE, (I)METHODRESPONSE children
   2686 parseEXPPARAMVALUE allows 2+ children, prohibits 0
   2685 Element.getAttribute returns empty string if no attribute
   2684 parseEXPMETHODRESPONSE has several issues
   2683 KEYVALUE VALUETYPE optional, "string" default
   2682 (I)MethodCallNode allows no LOCAL*PATH
   2681 parseQUALIFIERDECLARATION does not require TYPE attribute
   2680 IPARAMVALUE parsing broken on DOM/SAX
   2679 parseIMETHODCALL requires one IPARAMVALUE child element
   2678 parseMULTI___ allows one SIMPLE___ child element
   2677 ObjectPathNode allows all child nodes
   2676 parseMULTI(EXP)REQ looking for wrong child elements
   2675 CIMXMLParseException messages should contain element name
   2674 Null pointer exception in CIMDateTime(String)
   2673 NameSpaceNode does not need testCompletness()
   2672 Remove SIMPLEREQACK support
   2671 Potential null pointer exception in parseERROR
   2670 NAME attribute not required by DOM parser
   2669 Potential null pointer exception in parseMESSAGE
   2668 Potential null pointer exception in parseCIM
   2666 CR12: Remove ENUMERATIONCONTEXT

Version 2.2.4
================
   2661 Update detailed release history HTML for 2.2.4
   2662 Need the specific SSLHandshakeException during the cim call
   2660 CIMObjectPath.equalsModelPath same as equals
   2657 Potential null pointer exception in handleConnection
   2151 gzip compression not supported
   2655 Content-length must be ignored when Transfer-encoding present
   2651 IOException when tracing the cimxml
   2654 Check jcc idle time with CIMOM keepalive timeout to avoid EOF
   2653 FVT: java.lang.ExceptionInInitializerError during static init
   2652 LogAndTraceBroker.setXmlTraceStream should not close previous stream
   2650 SLP opaque value handling incorrect
   2604 SAXException messages should contain node name
   2647 Add two ssl protocol properties for http server and client
   
Version 2.2.3
================
   2643 Update detailed release history HTML for 2.2.3
   2642 Seperate properties needed for cim client and listener to filter out ciphers
   2635 Slowloris DoS attack for CIM indication listener port
   2640 Multiple CDATA parsing broken in DOM parser
   2639 CDATA parsing broken in PULL parser
   2638 Do not build empty REFERENCECLASS
   2637 Add nested embedded instance builder test
   2636 Nested embedded instances cause CIMXMLParseException
   2632 Potential Null Point Exception in CIMDataType
   2605 SAX parser throws wrong exception
   2628 Limit size of LinkedList of CIMEvents to be dispatched

Version 2.2.2
================
   2624 Update detailed release history HTML for 2.2.2
   2618 Need to add property to disable weak cipher suites for the secure indication
   2623 Reflect SourceForge upgrade in documentation
   2616 Add new API WBEMClientSBLIM.sendIndication()
   2621 Not all chunked input has trailers
   2620 Chunked output broken
   2619 Host should contain port when not 5988/5989
   2615 Add Allura links to detailed release history HTML
   2614 Remove redundant code in transmitRequest
3602604 Clean up SAXException messages
3601894 Enhance HTTP and CIM-XML tracing
3598613 different data type in cim instance and cim object path
3596303 windows http response WWW-Authenticate: Negotiate fails

Version 2.2.1
================
3584119 Update detailed release history HTML for 2.2.1
3592502 Enhance CIMDataType unit test
3588558 An enhancement on Java CIM Client logging
3557283 Print full response when get EOF from CIMOM
3576396 Improve logging of config file name
3572993 parseDouble("2.2250738585072012e-308") DoS vulnerability
3567433 Add links to top of detailed release history HTML

Version 2.2.0
================
3567429 Update detailed release history HTML for 2.2.0
3555752 Sync up javax.* javadoc with JSR48 1.0.0 Final IV
3565581 TCK: remove unnecessary overriding methods
3562235 LogAndTraceBrokerTest breaks unit test tracing

utst/org/sblim/cimclient/unittest/slp/SLPTest.java
/**
 * (C) Copyright IBM Corp. 2007, 2010
 *
 * THIS FILE IS PROVIDED UNDER THE TERMS OF THE ECLIPSE PUBLIC LICENSE 
 * ("AGREEMENT"). ANY USE, REPRODUCTION OR DISTRIBUTION OF THIS FILE 
 * CONSTITUTES RECIPIENTS ACCEPTANCE OF THE AGREEMENT.
 *
 * You can obtain a current copy of the Eclipse Public License from
 * http://www.opensource.org/licenses/eclipse-1.0.php
 *
 * @author : Endre Bak, IBM, ebak@de.ibm.com
 * 
 * Change History
 * Flag       Date        Prog         Description
 *------------------------------------------------------------------------------- 
 * 1804402    2007-09-28  ebak         IPv6 ready SLP
 * 1892103    2008-02-15  ebak         SLP improvements
 * 1949918    2008-04-08  raman_arora  Malformed service URL crashes SLP discovery
 * 2003590    2008-06-30  blaschke-oss Change licensing from CPL to EPL
 * 2204488 	  2008-10-28  raman_arora  Fix code to remove compiler warnings
 * 2524131    2009-01-21  raman_arora  Upgrade client to JDK 1.5 (Phase 1)
 * 2531371    2009-02-10  raman_arora  Upgrade client to JDK 1.5 (Phase 2)
 * 2823494    2009-08-03  rgummada     Change Boolean constructor to static
 * 3019214    2010-06-21  blaschke-oss SLP equals methods assume too much
 */

package org.sblim.cimclient.unittest.slp;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Enumeration;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.NoSuchElementException;
import java.util.Vector;
import java.util.logging.Level;

import org.sblim.cimclient.unittest.GenericUTestExts;
import org.sblim.cimclient.unittest.TestCase;
import org.sblim.slp.Advertiser;
import org.sblim.slp.Locator;
import org.sblim.slp.ServiceLocationAttribute;
import org.sblim.slp.ServiceLocationException;
import org.sblim.slp.ServiceLocationManager;
import org.sblim.slp.ServiceType;
import org.sblim.slp.ServiceURL;
import org.sblim.slp.internal.SLPConfig;
import org.sblim.slp.internal.TRC;
import org.sblim.slp.internal.sa.ServiceAgent;
import org.sblim.slp.internal.ua.SLEnumerationImpl;

/**
 * SLPTest
 * 
 */
public class SLPTest extends TestCase {

	/*
	 * Note: advertiser doesn't provide facility for scoped registrations so
	 * "default" is used always.
	 */

	private static class SrvEntry {

		private ServiceURL iSrvURL;

		private List<ServiceLocationAttribute> iAttribs;

		/**
		 * Ctor.
		 * 
		 * @param pSrvURL
		 * @param pAttribs
		 */
		public SrvEntry(String pSrvURL, ServiceLocationAttribute[] pAttribs) {
			this(new ServiceURL(pSrvURL, ServiceURL.LIFETIME_DEFAULT), pAttribs == null ? null
					: Arrays.asList(pAttribs));
		}

		/**
		 * Ctor.
		 * 
		 * @param pSrvURL
		 * @param pAttribs
		 */
		public SrvEntry(ServiceURL pSrvURL, List<ServiceLocationAttribute> pAttribs) {
			this.iSrvURL = pSrvURL;
			this.iAttribs = pAttribs;
		}

		/**
		 * getSrvURL
		 * 
		 * @return ServiceURL
		 */
		public ServiceURL getSrvURL() {
			return this.iSrvURL;
		}

		/**
		 * getAttribs
		 * 
		 * @return List
		 */
		public List<ServiceLocationAttribute> getAttribs() {
			return this.iAttribs;
		}

		/**
		 * register
		 * 
		 * @param pAdv
		 * @throws ServiceLocationException
		 */
		public void register(Advertiser pAdv) throws ServiceLocationException {
			pAdv.register(this.iSrvURL, this.iAttribs == null ? null
					: new Vector<ServiceLocationAttribute>(this.iAttribs));
		}

		/**
		 * deregister
		 * 
		 * @param pAdv
		 * @throws ServiceLocationException
		 */
		public void deregister(Advertiser pAdv) throws ServiceLocationException {
			pAdv.deregister(this.iSrvURL);
		}

		@Override
		public boolean equals(Object pObj) {
			if (pObj == null || !(pObj instanceof SrvEntry)) return false;
			SrvEntry that = (SrvEntry) pObj;
			return this.iSrvURL.equals(that.iSrvURL) && equalLists(this.iAttribs, that.iAttribs);
		}

	}

	private static final ServiceLocationAttribute[] ATTRIBS = new ServiceLocationAttribute[] {
			mkAttr("Description", new Object[] { "Lofasz", new byte[] { 0, 1, 2, 3 } }),
			mkAttr("Description", new Object[] { "ssh server", new Integer(10), Boolean.FALSE }),
			mkAttr("Description", new Object[] { "Nem lofasz" }) };

	private static final SrvEntry[] WBEM_SRVS = new SrvEntry[] {
			new SrvEntry("service:wbem://thishost.com:345",
					new ServiceLocationAttribute[] { ATTRIBS[2] }),
			new SrvEntry("service:wbem://thathost.com:3395",
					new ServiceLocationAttribute[] { ATTRIBS[0] }) };

	private static final SrvEntry[] SSH_SRVS = new SrvEntry[] {
			new SrvEntry("service:ssh://host.a.org:22", null),
			new SrvEntry("service:ssh://host.b.org:22",
					new ServiceLocationAttribute[] { ATTRIBS[1] }) };

	private static final ServiceType[] SRV_TYPES = new ServiceType[] {
			new ServiceType("service:wbem"), new ServiceType("service:ssh") };

	static boolean equalLists(List<?> pList0, List<?> pList1) {
		if (pList0 == null) return pList1 == null;
		if (pList1 == null) return false;
		if (pList0.size() != pList1.size()) return false;
		Iterator<?> itr0 = pList0.iterator();
		while (itr0.hasNext())
			if (!pList1.contains(itr0.next())) return false;
		return true;
	}

	private static ServiceLocationAttribute mkAttr(String pID, Object[] pValues) {
		return new ServiceLocationAttribute(pID, GenericUTestExts.mkVec(pValues));
	}

	private static List<ServiceURL> getURLs(SrvEntry[] pEntries) {
		List<ServiceURL> list = new ArrayList<ServiceURL>();
		for (int i = 0; i < pEntries.length; i++)
			list.add(pEntries[i].getSrvURL());
		return list;
	}

	private void checkEnum(Enumeration<?> pEnum, Object[] pRefObjs) {
		checkEnum(pEnum, Arrays.asList(pRefObjs));
	}

	private void checkEnum(Enumeration<?> pEnum, List<?> pRefList) {
		List<Object> resList = new ArrayList<Object>();
		while (pEnum.hasMoreElements())
			resList.add(pEnum.nextElement());
		verify("result count:" + resList.size() + " != reference count:" + pRefList.size()
				+ "\nresultList:" + resList, resList.size() == pRefList.size());
		Iterator<Object> resItr = resList.iterator();
		Iterator<?> refItr = pRefList.iterator();
		while (resItr.hasNext()) {
			Object resObj = resItr.next();
			verify("pRefList doesn't contain " + resObj, pRefList.contains(resObj));
		}
		while (refItr.hasNext()) {
			Object refObj = refItr.next();
			verify("resList doesn't contain " + refObj, resList.contains(refObj));
		}
	}

	private void checkSrvRequest(String pSrvType, SrvEntry[] pRefEntries)
			throws ServiceLocationException {
		debug("srvType:" + pSrvType);
		List<ServiceURL> srvList = getURLs(pRefEntries);
		Locator locator = ServiceLocationManager.getLocator(Locale.US);
		Enumeration<?> srvEnum = locator.findServices(new ServiceType(pSrvType), null, null);
		checkEnum(srvEnum, srvList);
	}

	private void checkSrvTypeRequest() throws ServiceLocationException {
		Locator locator = ServiceLocationManager.getLocator(Locale.US);
		Enumeration<?> srvTypeEnum = locator.findServiceTypes("", null);
		checkEnum(srvTypeEnum, SRV_TYPES);
	}

	private Vector<String> mkAttrIDVec(ServiceLocationAttribute[] pAttribs) {
		if (pAttribs == null) return null;
		Vector<String> vec = new Vector<String>();
		for (int i = 0; i < pAttribs.length; i++)
			vec.add(pAttribs[i].getId());
		return vec;
	}

	private void checkAttribRequest(String pURLStr, ServiceLocationAttribute[] pAttribs)
			throws ServiceLocationException {
		Locator locator = ServiceLocationManager.getLocator(Locale.US);
		Enumeration<?> attrEnum = locator.findAttributes(new ServiceURL(pURLStr,
				ServiceURL.LIFETIME_DEFAULT), null, mkAttrIDVec(pAttribs));
		checkEnum(attrEnum, pAttribs);
	}

	private void checkAttribRequest(ServiceType pSrvType, ServiceLocationAttribute[] pAttribs)
			throws ServiceLocationException {
		Locator locator = ServiceLocationManager.getLocator(Locale.US);
		Enumeration<?> attrEnum = locator.findAttributes(pSrvType, null, mkAttrIDVec(pAttribs));
		checkEnum(attrEnum, pAttribs);
	}

	private ServiceAgent iAgent;

	private static final int PORT = 11223;

	private void init() throws IOException {
		if (this.iAgent != null) return;
		TRC.setLevel(Level.SEVERE);
		TRC.setPatterns(new String[] {
		// "^.+$"
				}, new String[] { "^.*\\.SLPTest\\..+$", "^.*\\.MsgHeader\\..+$"
				// "^.*\\.Requester\\..+$",
				});
		TRC.debug("test");
		SLPConfig.getGlobalCfg().setPort(PORT);
		// SLPConfig.getGlobalCfg().setUseIPv4(false); // Windows won't like it
		// SLPConfig.getGlobalCfg().setUseIPv6(false);
		this.iAgent = new ServiceAgent();
		this.iAgent.start();
	}

	/**
	 * testReqistration
	 * 
	 * @throws Exception
	 */
	public void testReqistration() throws Exception {
		try {
			init();
			this.iAgent.setSkipFirstRequest(true);
			Advertiser adv = ServiceLocationManager.getAdvertiser(Locale.US);
			// registration
			for (int i = 0; i < WBEM_SRVS.length; i++)
				WBEM_SRVS[i].register(adv);
			for (int i = 0; i < SSH_SRVS.length; i++)
				SSH_SRVS[i].register(adv);
		} catch (Exception e) {
			this.iAgent.stop();
			throw e;
		} finally {
			this.iAgent.setSkipFirstRequest(false);
		}
	}

	/**
	 * testSrvRequest
	 * 
	 * @throws Exception
	 */
	public void testSrvRequest() throws Exception {
		try {
			init();
			checkSrvRequest("service:wbem", WBEM_SRVS);
			checkSrvRequest("service:ssh", SSH_SRVS);
		} catch (Exception e) {
			this.iAgent.stop();
			throw e;
		}
	}

	/**
	 * testDeregistration
	 * 
	 * @throws Exception
	 */
	public void testDeregistration() throws Exception {
		try {
			init();
			Advertiser adv = ServiceLocationManager.getAdvertiser(Locale.US);
			SSH_SRVS[0].deregister(adv);
			checkSrvRequest("service:ssh", new SrvEntry[] { SSH_SRVS[1] });
		} catch (Exception e) {
			this.iAgent.stop();
			throw e;
		}
	}

	/**
	 * testSrvTypeRequest
	 * 
	 * @throws Exception
	 */
	public void testSrvTypeRequest() throws Exception {
		try {
			init();
			checkSrvTypeRequest();
		} catch (Exception e) {
			this.iAgent.stop();
			throw e;
		}
	}

	/**
	 * testAttrRequest
	 * 
	 * @throws Exception
	 */
	public void testAttrRequest() throws Exception {
		try {
			init();
			checkAttribRequest("service:wbem://thathost.com:3395",
					new ServiceLocationAttribute[] { ATTRIBS[0] });
			checkAttribRequest(new ServiceType("service:wbem"), new ServiceLocationAttribute[] {
					ATTRIBS[0], ATTRIBS[2] });
		} catch (Exception e) {
			this.iAgent.stop();
			throw e;
		}
	}

	/**
	 * testOverflowHandling
	 * 
	 * @throws Exception
	 */
	public void testOverflowHandling() throws Exception {
		try {
			init();
			Advertiser adv = ServiceLocationManager.getAdvertiser(Locale.US);
			SrvEntry[] entries = new SrvEntry[128];
			for (int i = 0; i < entries.length; i++) {
				entries[i] = new SrvEntry("service:https://dummy." + i + ".org:443",
						new ServiceLocationAttribute[] { mkAttr("ID",
								new Object[] { new Integer(i) }) });
				entries[i].register(adv);
			}
			checkSrvRequest("service:https", entries);
		} catch (Exception e) {
			this.iAgent.stop();
			throw e;
		}
	}

	/**
	 * testExceptionHandling
	 * 
	 * @throws Exception
	 */
	public void testExceptionHandling() throws Exception {
		Locator locator = ServiceLocationManager.getLocator(Locale.US);
		Enumeration<?> srvEnum = locator.findServices(new ServiceType("service:wbem"), null, null);
		// WBEM_SRVS.length number of services should be read
		for (int i = 0; i < WBEM_SRVS.length; i++) {
			verify("One more element is expected", srvEnum.hasMoreElements());
			Object obj = srvEnum.nextElement();
			String typeStr = obj == null ? "null" : obj.getClass().getName();
			verify("obj is " + typeStr, obj instanceof ServiceURL);
		}
		/*
		 * Now let's send a garbage packet to the UA, as a result nextElement()
		 * should not throw a RuntimeException with a parsing error cause.
		 * Parsing error saves exception into ExceptionTable
		 */
		DatagramSocket socket = new DatagramSocket();
		byte[] reqBytes = new byte[] { (byte) 0xff, 0, 0 };
		int port = ((SLEnumerationImpl) srvEnum).getPort();
		DatagramPacket outPacket = new DatagramPacket(reqBytes, reqBytes.length, InetAddress
				.getByName("127.0.0.1"), port);
		socket.send(outPacket);

		verify("One more element is expected", !srvEnum.hasMoreElements());
		try {
			srvEnum.nextElement();
			verify("Exception was not thrown!", false);
		} catch (Exception e) {
			String cause = e.getCause() == null ? "null" : e.getCause().toString();
			verify("Wrong cause: " + cause, e instanceof NoSuchElementException);
		}
		verify("Unexpected element.", !srvEnum.hasMoreElements());
	}

	/**
	 * testStop
	 */
	public void testStop() {
		if (this.iAgent != null) this.iAgent.stop();
	}

	/**
	 * @param pMsg
	 */
	private static void debug(String pMsg) {
	// System.out.println("DEBUG:"+pMsg);
	}

}
